### ISSUE TITLE
Some parts are missing when using versions 5.1.xxx

### ISSUE THREAD

**Issue by KevinFournier** [xBimTeam/XbimGeometry#261]
Hello,

I have encountered an issue when trying to reconstruct the geometry from an ifc; some parts of the model are missing.

**Context:** I am making a .Net application to send the data of an ifc file through the network and read it in another application (Unity3D) to display the model and information about it.
I do not know the process of creation of the ifc files I work with and the files could come from any Bim related software.

**What I have done:** I tried to visualize the ifc file with XbimXplorer:
 - With the version downloaded on your website, I did not find any part missing.
 - With a build of the XbimWindowsUI project on my computer, some part went missing again.

The parts missing when I use the data send over the network and the one on the built version of XbimXplorer are the same.
Here are some images to illustrate the previous explanation and the assemblies use to build XbimXplorer on my computer:

- Downloaded XbimXplorer:
![xBimXplorer_exe](https://user-images.githubusercontent.com/11573588/79993107-ad9dc880-84b4-11ea-8eef-1359bf2a59b8.PNG)

- Built XbimXplorer:
![xBimXplorer_build](https://user-images.githubusercontent.com/11573588/79993108-ae365f00-84b4-11ea-95a9-8e89fb6e9286.PNG)

- Assembly details of the built version:
![xBimXplorer_build_Assembly_Details](https://user-images.githubusercontent.com/11573588/79993105-ad053200-84b4-11ea-9455-3a82ebf8c538.PNG)

I have found a workaround to my problem by using the version 5.0 of the Xbim NuGet packages (essential and geometry) instead of the version 5.1

I currently use those packages:
- Common - 5.0.213
- Ifc - 5.0.213
- Ifc2x3 - 5.0.213
- Ifc4 - 5.0.213
- IO.MemoryModel - 5.0.213
- Tessellator - 5.0.213
- Geomtery - 5.0.163
- Geometry.Engine.Interop - 5.0.163
- ModelGeometry.Scene - 5.0.163

When I update to version 5.1.xxx, the missing parts are missing again.

The error may still be on my side because I am new to Xbim (and bim) and might have done something wrong when reading the geometry data. I am sorry for the disarrangement if that's the case.

Regards,

Kévin Fournier

**Comment by andyward** [comment]
Are there any warnings generated? Xplorer will show a small icon in the bottom toolbar if there are any. Clicking this will show the logs

**Comment by KevinFournier** [comment]
Andy, 

Yes, there is errors in the "old" version of xBimExplorer but they do not seem to impact the geometry.
In the build version, however, there is only warning but the geometry is affected.
Here is few images:

**Build version, with warning, missing geometry:**
![xBimXplorer_Build_Logs](https://user-images.githubusercontent.com/11573588/80191643-4861e880-8616-11ea-8476-d3c191a0f32f.PNG)

**Exe version, with errors**
![xBimXplorer_exe_Logs](https://user-images.githubusercontent.com/11573588/80191648-48fa7f00-8616-11ea-9ba4-1b35a6114672.PNG)

----

Steve,

I can provide you the file I used in the previous example.

[Bossage_P08.zip](https://github.com/xBimTeam/XbimGeometry/files/4527635/Bossage_P08.zip)


---

I tried with another file (that I can’t show nor share, unfortunately). In this one, there is also missing geometry when I reconstruct the geometry in Unity but it doesn’t seem to have missing geometry in both xBimVersion I used.
This file is much bigger, however, and some parts could be missing without notice (There is obvious missing of geometry part in my read and reconstruct app).

I am waiting for more file to be given to me for further testing. I should have more information by the mid of next week.

Kévin F.

**Comment by SteveLockley** [comment]
Thanks Kevin I will have a look

**Comment by SteveLockley** [comment]
Kevin I have found and fixed the problem with this model. Its a good exposure of a quite difficult edge case. Thanks for taking the time to publish it. Basically there are three issues that caused this omission of geometry, they are all in IfcCompositeCurve #1315.
1) The model you have is not respecting the tolerances specific in the representation context, specifically the Composite Curve eighth segment  is 2.3395e-4 from the end of the previous segment (seventh), the model precision is 1e-5. This may seem small but it happens to make a great difference in the next condition
2) The composite curve is self intersecting, the first segment wholly overlaps with 7th segment, whilst legal for a composite curve it is not legal for the swept solid it represents. This plus the tolerance issue meant that xbim found the nearest point to be the end of the first segment and as this resulted in a non-manifold wire it gave up on the rest.
3) Finally, the xbim wire builder used OCC wire maker which by default when adding segments to a wire hooks the start or end of segment to any nearest point of the rest of the composite curve, Hence the wire was not contiguous or manifold. I have now changed the code so that it gives priority to the start or end of the composite curve, not just the nearest.
![image](https://user-images.githubusercontent.com/8222629/80702412-ebb07300-8ad8-11ea-8819-5c25a50dc44b.png)
